### PR TITLE
interop-testing: New client for HTTP/2 "negative" interop tests

### PR DESCRIPTION
**Pull request by ericgribkoff**
These tests are for the new Http/2 test server (https://github.com/grpc/grpc/pull/8900). They are designed to test client behavior when the server sends unexpected responses during rpcs.

@ncteisen @makdharma FYI

### CODE REVIEW

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+    client.parseArgs(args);
+    client.setUp();
+
+    Runtime.getRuntime().addShutdownHook(new Thread() {
+      @Override
+      public void run() {
+        try {
+          client.tearDown();
+        } catch (Exception e) {
+          e.printStackTrace();
```
Comment (carl-mastrangelo) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: should be done using a logger.

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        }
+      }
+    });
+
+    try {
+      client.run();
+    } finally {
+      client.tearDown();
+    }
+    System.exit(0);
```
Comment (carl-mastrangelo) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: no need for this.  It should be implicit.

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+  }
+
+  private String serverHost = "localhost";
+  private int serverPort = 8080;
+  private String testCase = "rst_stream_after_header";
```
Comment (carl-mastrangelo) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: This doesn't match any of the enums.

Better to do something like `RST_AFTER_HEADER.name()`

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+    blockingStub = TestServiceGrpc.newBlockingStub(channel);
+    asyncStub = TestServiceGrpc.newStub(channel);
+  }
+
+  private synchronized void tearDown() {
+    try {
+      if (channel != null) {
+        channel.shutdown();
+      }
+    } catch (RuntimeException ex) {
```
Comment (carl-mastrangelo) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: no need to catch RuntimeException

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+    }
+  }
+
+  private void setUp() {
+    channel = createChannel();
+    blockingStub = TestServiceGrpc.newBlockingStub(channel);
+    asyncStub = TestServiceGrpc.newStub(channel);
+  }
+
+  private synchronized void tearDown() {
```
Comment (carl-mastrangelo) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: why synchronized?

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+      }
+    } catch (RuntimeException ex) {
+      throw ex;
+    } catch (Exception ex) {
+      throw new RuntimeException(ex);
+    }
+  }
+
+  private void run() {
+    System.out.println("Running test " + testCase);
```
Comment (carl-mastrangelo) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: This too should be a log statement

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+        throw new IllegalArgumentException("Unknown test case: " + testCase);
+    }
+  }
+
+  private class Tester {
+    private final SimpleRequest simpleRequest = SimpleRequest.newBuilder()
+        .setResponseSize(314159)
+        .setResponseType(PayloadType.COMPRESSABLE)
+        .setPayload(Payload.newBuilder()
+            .setBody(ByteString.copyFrom(new byte[271828])))
```
Comment (carl-mastrangelo) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: does this fit on the previous line?

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+        .build();
+
+    public void rstStreamAfterHeader() throws Exception {
+      try {
+        blockingStub.unaryCall(simpleRequest);
+        fail("Expected call to fail");
+      } catch (StatusRuntimeException ex) {
+        assertEquals(Status.Code.UNAVAILABLE, ex.getStatus().getCode());
```
Comment (carl-mastrangelo) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: you should also be checking for a substring in the error message.  There are several cases where you check for UNAVAILABLE, but this doesn't distinguish them.

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+    public void rstStreamAfterData() throws Exception {
+      try {
+        blockingStub.unaryCall(simpleRequest);
+        fail("Expected call to fail");
+      } catch (StatusRuntimeException ex) {
+        assertEquals(Status.Code.UNAVAILABLE, ex.getStatus().getCode());
+      }
+
+      // Use async stub to verify data was received.
+      StreamRecorder<SimpleResponse> responseObserver = StreamRecorder.create();
```
Comment (carl-mastrangelo) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: @ejona86  declined PR #2385 on the grounds that StreamRecoder was going to be deleted.  

You probably don't want to use it here.

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+      responseObserver.awaitCompletion();
+      assertEquals(1, responseObserver.getValues().size());
+      assertEquals(Status.Code.UNAVAILABLE,
+          Status.fromThrowable(responseObserver.getError()).getCode());
+    }
+
+    public void rstStreamDuringData() throws Exception {
+      try {
+        blockingStub.unaryCall(simpleRequest);
+        fail("Expected call to fail");
```
Comment (carl-mastrangelo) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: I'm very curious why JUnit methods are being used in a non JUnit environment.  Can these be avoided?

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+      assertEquals(Status.Code.UNAVAILABLE,
+          Status.fromThrowable(responseObserver.getError()).getCode());
+    }
+
+    public void goAway() throws Exception {
+      blockingStub.unaryCall(simpleRequest);
+      blockingStub.unaryCall(simpleRequest);
+    }
+
+    public void ping() throws Exception {
```
Comment (carl-mastrangelo) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: why public?

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+
+      ListeningExecutorService threadpool =
+          MoreExecutors.listeningDecorator(newFixedThreadPool(numThreads));
+      List<ListenableFuture<?>> workerFutures = new ArrayList<ListenableFuture<?>>();
+      ManagedChannel channel = createChannel();
+      for (int i = 0; i < numThreads; i++) {
+        workerFutures.add(threadpool.submit(new MaxStreamsWorker(channel, i, simpleRequest)));
+      }
+      ListenableFuture<?> f = Futures.allAsList(workerFutures);
+      f.get();
```
Comment (carl-mastrangelo) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: Warning, if one of the futures fails, the rest of them will not be forced.  (resolved?)  Does this code cope with some futures not completing?

**Comment by ericgribkoff** [comment]
@carl-mastrangelo I pushed a commit that addresses (hopefully) all of your review comments, including dealing with failed futures in the maxStreams test.

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+      default:
+        throw new IllegalArgumentException("Unknown test case: " + testCase);
+    }
+  }
+
+  private class Tester {
+    private final int timeoutSeconds = 5;
+
+    private final SimpleRequest simpleRequest = SimpleRequest.newBuilder()
+        .setResponseSize(314159)
```
Comment (adelez) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: Define constants for these sizes?

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+        .setResponseType(PayloadType.COMPRESSABLE)
+        .setPayload(Payload.newBuilder().setBody(ByteString.copyFrom(new byte[271828])))
```
Comment (adelez) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: Same here.

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+        throw new AssertionError("Operation timed out");
+      }
+      if (responseObserver.getResponses().size() != 0) {
+        throw new AssertionError("Expected zero responses");
+      }
+      assertRstStreamReceived(Status.fromThrowable(responseObserver.getError()));
+    }
+
+    private void goAway() throws Exception {
+      blockingStub.unaryCall(simpleRequest);
```
Comment (adelez) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: Should we add some verification for the cases in which the calls succeed?

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java (hunk 13 above)
Comment (ericgribkoff) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: Done.

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java (hunk 14 above)
Comment (ericgribkoff) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: Done.

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java (hunk 15 above)
Comment (ericgribkoff) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: Done.

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java [hunk 16]
```diff
@@ -0,0 +1,401 @@
...
+    final SimpleResponse goldenResponse = SimpleResponse.newBuilder()
+        .setPayload(Payload.newBuilder()
+            .setType(PayloadType.COMPRESSABLE)
+            .setBody(ByteString.copyFrom(new byte[responseSize])))
+        .build();
+
+    private void rstAfterHeader() throws Exception {
+      try {
+        blockingStub.unaryCall(simpleRequest);
+        throw new Exception("Expected call to fail");
```
Comment (carl-mastrangelo) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: AssertionError

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java [hunk 17]
```diff
@@ -0,0 +1,401 @@
...
+
+  private ManagedChannel createChannel() {
+    InetAddress address;
+    try {
+      address = InetAddress.getByName(serverHost);
+    } catch (UnknownHostException ex) {
+      throw new RuntimeException(ex);
+    }
+    return NettyChannelBuilder.forAddress(new InetSocketAddress(address, serverPort))
+        .flowControlWindow(65 * 1024)
```
Comment (carl-mastrangelo) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: why mess with the Flow control window?

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java (hunk 16 above)
Comment (ericgribkoff) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: Done

File: interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java (hunk 17 above)
Comment (ericgribkoff) [inline interop-testing/src/main/java/io/grpc/testing/integration/Http2Client.java]: No reason. Removed this call.

**Pull request merged**